modules/dag-l1/src/main/scala/org/tessellation/dag/l1/StateChannel.scala

```diff
@@ -177,16 +190,21 @@ class StateChannel[F[_]: Async: KryoSerializer: SecurityProvider: Random](
         }
         .void
     )
+    .evalTap(_ => semaphores.blockAcceptance.release)
 
   private val globalSnapshotProcessing: Stream[F, Unit] = Stream
     .awakeEvery(10.seconds)
     .evalMap(_ => services.l0.pullGlobalSnapshots)
     .flatMap(snapshots => Stream.fromIterator(snapshots.iterator, 1))
-    .evalMap { snapshot =>
-      snapshot.hashF.flatMap { hash =>
-        logger.info(s"Pulled following global snapshot: ${snapshot.ordinal.show} ${hash.show}")
-      }
-    }
+    .evalTap(snapshot => logger.info(s"Pulled following global snapshot: $snapshot"))
+    .evalTap(
+      _ => semaphores.blockAcceptance.acquire >> semaphores.blockCreation.acquire >> semaphores.blockStoring.acquire
+    )
+    .evalMap(programs.snapshotProcessor.process)
+    .evalTap(
+      _ => semaphores.blockAcceptance.release >> semaphores.blockCreation.release >> semaphores.blockStoring.release
+    )
```

Comment:
if the only reason to use semaphore is to pause acceptance/creation/storage in that place I'd recommend using a `SignallingRef` and pausing above streams instead of locking 🤔

Reply:
Makes sense as the requirement I have is a single permit semaphore. SignalingRef makes sense for that use.

Reply:
Looked into pausing the streams but it actually doesn't give me the same guarantees as using semaphores. Doesn't matter where I would put `pauseWhen` (I guess that's what was meant to be used here in conjunction with `SignalingRef`) the stream will pause on the next iteration or rather after reaching the `pauseWhen` clause for the next time. Hence storing a block after setting the `signalingRef` to `false` may still happen - so we get a race condition. Whereas when using semaphores we acquire a permit before performing an action and nothing else can use that semaphore (true when using single permit semaphores like I do here) in the meantime, so I have a guarantee that the state of the storage will not be changed. Correct me if I'm wrong but it seems to me that semaphores are still the best way to go.